apps/extension/src/pages/formatting.tsx

```diff
@@ -0,0 +1,287 @@
+import { DateFormat, IntroSentenceType } from "@agreeto/api/types";
+import {
+  DEFAULT_LANGUAGE_FORMAT,
+  LANGUAGE_FORMATS,
+  type LanguageFormatItem,
+  extractTextFromSlots,
+} from "@agreeto/calendar-core";
+import * as DropdownMenu from "@radix-ui/react-dropdown-menu";
+import { addHours, format, startOfDay } from "date-fns";
+import { useEffect, useRef, useState } from "react";
+import { HiCheckCircle } from "react-icons/hi";
+import { TiArrowSortedDown } from "react-icons/ti";
+
+import { trpcApi } from "~features/trpc/api/hooks";
+
+// @ts-expect-error - svg's are not typed
+import previewImage from "./../../assets/gmail-preview.svg";
+
+export const Formatting = () => {
+  const utils = trpcApi.useContext();
+  const introSentenceRef = useRef<HTMLTextAreaElement>(null);
+  const [languageDropdownOpen, setLanguageDropdownOpen] = useState(false);
+  const [dateFormatDropdownOpen, setDateFormatDropdownOpen] = useState(false);
+  const [introSentenceFocused, setIntroSentenceFocused] = useState(false);
+  const [canSave, setCanSave] = useState(false);
+  const [introSentence, setIntroSentence] = useState<string>(
+    DEFAULT_LANGUAGE_FORMAT.defaultIntroSentence,
+  );
+  const [introSentenceType, setIntroSentenceType] = useState<IntroSentenceType>(
+    IntroSentenceType.DEFAULT,
+  );
+  const [selectedLanguage, setSelectedLanguage] = useState<LanguageFormatItem>(
+    DEFAULT_LANGUAGE_FORMAT,
+  );
+  const [selectedDateFormat, setSelectedDateFormat] = useState<DateFormat>(
+    DateFormat.MMMM_d_EEEE,
+  );
+
+  const { data: formattings } = trpcApi.formatting.byCurrentUser.useQuery();
+  const { mutate: updateFormatting, isLoading: isUpdating } =
+    trpcApi.formatting.update.useMutation({
+      onSuccess() {
+        utils.formatting.byCurrentUser.invalidate();
+      },
+    });
+
+  useEffect(() => {
+    const formatting = formattings?.find(
+      (f) => f.language === selectedLanguage.key,
+    );
+
+    if (formatting) {
+      setSelectedDateFormat(formatting.dateFormat);
+      setSelectedLanguage(
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        LANGUAGE_FORMATS.find((i) => i.key === formatting.language)!,
+      );
+      setIntroSentenceType(formatting.introSentenceType);
+      setIntroSentence(formatting.introSentence);
+    } else {
+      // Reset fields
+      setSelectedDateFormat(DateFormat.MMMM_d_EEEE);
+      setIntroSentenceType(IntroSentenceType.DEFAULT);
+      resetIntroSentence(IntroSentenceType.DEFAULT);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [selectedLanguage, formattings]);
```

Comment:
This is a bit sketchy imo, no need to make all of this state and synchronizing in an effect. We can inherit these values from the formatting as regular variables

It's sort of similar to the antipattern:
```ts
const [firstName, setFirstName] = useState("")
const [lastName, setLastName] = useState("")
const [fullName, setFullName] = useState(firstName + lastName)

useEffect(() => {
  setFullName(firstName + lastName)
}, [firstName, lastName])
```

where you could just 
```ts
const [firstName, setFirstName] = useState("")
const [lastName, setLastName] = useState("")
const fullName = firstName + lastName
```

Reply:
Yeah would be cool if we can make this part more clear. But what do you suggest? I didn't get it.

Reply:
That's a great discussion here that adds a lot of value to our engineering. I appreciate both comments. 

The tl;dr of my comment:
- too many useState/useEffect's are a sign of handling react state inefficiently leading to bugs and increased maintainability
- given how fundamental our dependency on react is & best practices being readily available, we should spend the time to fix this now and not some time in the future

And here's the actual content 👇

I agree that the usage of these useEffect to sync states is a code smell to me. 

I'm also not sure specifically what you're referring to @juliusmarminge, but the best reference on deciding to useEffect or not for me are Dan abramov's docs on it, which are the official react beta docs. 

Eg, it looks like we could handle the state in an event handler instead.

![image](https://user-images.githubusercontent.com/18185649/205007776-08d6f24f-f6d4-4107-bfa4-dc39fecbbcee.png)


Another point that's stated is to not useState at all, where possible thereby passing the need to sync state between each other. 
![image](https://user-images.githubusercontent.com/18185649/205008559-3bf5cc46-074d-42c4-a22c-79bb151dddfb.png)
![image](https://user-images.githubusercontent.com/18185649/205008578-20789cdf-3c1e-4f0a-8eba-58f38d02f3a6.png)

Lastly, I think we should also not reset state based on prop changes if we can avoid that, as per the docs:
![image](https://user-images.githubusercontent.com/18185649/205008873-bdde617f-5a7e-48b5-a9f3-b1e16fa5c99d.png)
![image](https://user-images.githubusercontent.com/18185649/205009326-8690eed1-a1a7-4088-b013-eb67760923cf.png)

![image](https://user-images.githubusercontent.com/18185649/205009650-7342960f-bcc2-45b2-9215-902b204734a9.png)

For form data specifically, I think state is overused. It's possible to use the web browser's API for this and remix demonstrated this and some react form libraries take the same approach: use browser api's over custom and unnecessary react abstractions. The big advantage is that react is based on JS with this in mind, so it plays together quite nicely. 

Eg consider this form that doesn't event need a hook nor a form library:
![image](https://user-images.githubusercontent.com/18185649/205013448-97193772-41d9-46f8-a628-4fcf6e5923ab.jpeg)
![image](https://user-images.githubusercontent.com/18185649/205013532-c9967b2f-cdf8-491c-9f37-9939930993b1.jpeg)
[Source](https://twitter.com/DavidKPiano/status/1569361453928300545?s=20&t=ttjxrYx4pkH7F5YwaISFPQ) 


That's what I like about react-zorm:
It uses browser APIs where possible, but adds Typesafety where useful. An example for this is the event.target type, which is an issue in TypeScript with FormData. You can use zod to make it type safe, which is basically what react-zorm does for you as a lib conceptionally. 

All of these best practices are to use react's data flow for state more efficiently. 

While I'm aware that understanding react's state flow throughout the component cycle is hard to understand, I think making an effort to learn this pays off in the future from my personal experience. Adding state and useEffects leads to components which are very hard to debug because it's not always straightforward to understand the data flow of state in react components. Therefore, if we can keep it as vanilla and follow react's recommendations as much as possible, we will improve maintainability. 

If we do encounter questions about this in the future, I found that [Dan was also responsive](https://twitter.com/dan_abramov/status/1541821478009798658?s=20&t=eGg9MO7Wip7l3251o16ujg) into looking into specific use cases but I think here it just pays off to follow React's recommended practices for when not to useEffect. That should already clean up quite a bit.

Reply:
Having said that, I do believe that there's. Big value in tying our form library into our router. 

Remix has demonstrated how nicely this enables optimistic UI without the downsides. 

Given that trpc enables us to invalidate not only based on routes but other calls on top of that, I'd be happy with a form<>tRPC binding but would honestly prefer a router<>form binding that is put on steroids by using tRPC. 

That's my take.

Reply:
The name example was somewhere in those screenshots @p6l-richard posted.

Essentially, any data that is inherited from another state shouldn't be state in itself. In our case this would be the 
```ts
 const [introSentence, setIntroSentence] = useState<string>(
     DEFAULT_LANGUAGE_FORMAT.defaultIntroSentence,
   );
   const [introSentenceType, setIntroSentenceType] = useState<IntroSentenceType>(
     IntroSentenceType.DEFAULT,

useEffect(() => {
  // update when formatting changes
}, [formatting]);
```
to 
```ts
const introSentence = formatting?.introSentence ?? DEFAULT_LANGUAGE_FORMAT.defaultIntroSentence;
const introSentenceType = formatting?.introSentenceType ?? IntroSentenceType.DEFAULT
```

No need to synchronize them in an effect.

For stuff that still would need synchronization, the useQuery's onSuccess, onError handlers are better here in the sense that we use the primitives that our state-provider gives us.

I don't know if that made it clearer?